### PR TITLE
Port to 1.20.6

### PR DESCRIPTION
**Pull request by voidpointer0x00**
At this point I just made it possible to build the mod, haven't tested anything though. I will do _some_ testing, but the only important feature for me is the accurate block placement thingy — so I likely will not test much further than that.

I am most uncertain of the changes made to `src/main/java/carpetextra/utils/PlaceBlockDispenserBehavior.java`, see the diff.

Brief overview of the changes:
- BlockEntity data switched from raw nbt to data component types API;
- a different `#damage()` method for item stacks;
- dispensers behaviour mixin target changed, updated to use data component types;
- suspicious stew effects types were changed a little, shouldn't break though.

Before merge _someone_ should test:
- [ ] if dispensers related fatures still work
  - [x] blocks
  - [x] skulls
  - [x] chests
  - [x] carpets
  - [x] shulkers
  - [ ] anything else?
- [x] NetherWartBlock fertilization works
- [x] skeletons can be converted to wither skeletons with lightning
- [ ] https://github.com/voidpointer0x00/carpet-extra/commit/092ee7e758d8e14fa0751f749bfb2e690a0ecae0 crash does not occur in this version

Commits probably should be squashed when merged :grin: 

If everything is ok
- [ ] update carpet-extra version in `gradle.properties`

## Builds

https://github.com/voidpointer0x00/carpet-extra/releases

### CODE REVIEW

**Comment by voidpointer0x00** [comment]
Fixed more mixins, made it actually runnable :grin: 

Built mod can be found here
https://github.com/voidpointer0x00/carpet-extra/releases/tag/1.20.6-pre

File: src/main/java/carpetextra/mixins/DispenserBehaviorMobHeadMixin.java [hunk 1]
```diff
@@ -1,19 +1,17 @@
...
 import net.minecraft.block.dispenser.FallibleItemDispenserBehavior;
 import net.minecraft.item.BlockItem;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.math.BlockPointer;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
 
-@Mixin(targets = {"net/minecraft/block/dispenser/DispenserBehavior$14", "net/minecraft/block/dispenser/DispenserBehavior$15"})
```
Comment (MeeniMc) [inline src/main/java/carpetextra/mixins/DispenserBehaviorMobHeadMixin.java:14]: I was trying to do this port as well but you are way more advanced, but when I was doing it I had $14->$3 and $15->$4 but you have only $3

Comment (voidpointer0x00) [reply src/main/java/carpetextra/mixins/DispenserBehaviorMobHeadMixin.java:14]: I just assumed Mojang changed the original code, I'll check again and report here 😄

File: src/main/java/carpetextra/dispenser/behaviors/MilkMooshroomDispenserBehavior.java [hunk 2]
```diff
@@ -57,11 +56,11 @@ private static ItemStack getStewType(List<MooshroomEntity> mooshrooms) {
         for (MooshroomEntity mooshroom : mooshrooms) {
             // from MooshroomEntity#interact
             MooshroomEntity_StatusEffectAccessorMixin mooshroomAccessor = (MooshroomEntity_StatusEffectAccessorMixin) mooshroom;
-            List<StewEffect> stewEffects = mooshroomAccessor.getStewEffects();
+            var stewEffects = mooshroomAccessor.getStewEffects();
             if (stewEffects != null) {
                 // create suspicious stew and add mooshroom's stew effect to it
                 ItemStack stewStack = new ItemStack(Items.SUSPICIOUS_STEW);
-                SuspiciousStewItem.writeEffectsToStew(stewStack, stewEffects);
+                stewStack.set(DataComponentTypes.SUSPICIOUS_STEW_EFFECTS, mooshroomAccessor.getStewEffects());
```
Comment (MeeniMc) [inline src/main/java/carpetextra/dispenser/behaviors/MilkMooshroomDispenserBehavior.java:63]:
```suggestion
                stewStack.set(DataComponentTypes.SUSPICIOUS_STEW_EFFECTS, stewEffects);
```

File: src/main/java/carpetextra/mixins/DispenserBehaviorMobHeadMixin.java (hunk 1 above)
Comment (voidpointer0x00) [reply src/main/java/carpetextra/mixins/DispenserBehaviorMobHeadMixin.java:14]: On the latest 1c54c8cb commit `$14` and `$15` were mixing into dispensing glowstone and glass bottles, that's weird. Looked into the new code — the right bytecode for mob heads should be `$11` for mob heads and `$12` for wither skeleton skulls, `$3` is a saddle :smile: 

I guess I'm gonna have to just check whether if it works in game :grin:

Comment (voidpointer0x00) [reply src/main/java/carpetextra/mixins/DispenserBehaviorMobHeadMixin.java:14]: The `$3` is the one used with heads and skeleton skulls, I did miss the `$4` for wither skeleton skull. Thanks for your attention :)

Could not quite figure out why it worked anyway, found out the carpet dispenser was dispensing wither skulls :smile: Going to recheck all the targets and fix it.

Comment (voidpointer0x00) [reply src/main/java/carpetextra/mixins/DispenserBehaviorMobHeadMixin.java:14]: Fixed with recent commits

**Comment by voidpointer0x00** [comment]
Published release with chest and carpet dispenser fixes [1.20.6-pre2](https://github.com/voidpointer0x00/carpet-extra/releases/tag/1.20.6-pre2)

**Comment by gamma-02** [comment]
I've tested the three things it looks like still need testing on a server and on my client, all seems well!

File: src/main/java/carpetextra/mixins/DispenserBehaviorCarpetsMixin.java [hunk 3]
```diff
@@ -9,10 +9,9 @@
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
 
-@Mixin(targets = "net/minecraft/block/dispenser/DispenserBehavior$4")
+@Mixin(targets = "net/minecraft/block/dispenser/DispenserBehavior$13")
```
Comment (MeeniMc) [inline src/main/java/carpetextra/mixins/DispenserBehaviorCarpetsMixin.java:12]: not sure if defect, but placing carpets with carpet carpets rule on does not issue the result of the command (e.g., pink carpet should print the spawn conditions). I'll review prior versions behavior and mark resolved if not needed.

File: src/main/java/carpetextra/dispenser/behaviors/ToggleBlockDispenserBehavior.java [hunk 4]
```diff
@@ -52,7 +52,7 @@ protected ItemStack dispenseSilently(BlockPointer pointer, ItemStack stack) {
             BlockHitResult hitResult = new BlockHitResult(Vec3d.of(frontBlockPos), dispenserFacing.getOpposite(), frontBlockPos, false);
 
             // try to use block
-            if(frontBlockState.onUse(world, null, Hand.MAIN_HAND, hitResult).isAccepted()) {
+            if(frontBlockState.onUseWithItem(stack, world, null, Hand.MAIN_HAND, hitResult).isAccepted()) {
```
Comment (MeeniMc) [inline src/main/java/carpetextra/dispenser/behaviors/ToggleBlockDispenserBehavior.java:55]: broken?: this does nothing, the stick does not shoot out, but it doesn't toggle the sate. Not sure -what- this is supposed to do, as using this with buttons/levers would lock the dispenser? Maybe it was also broken in prior versions?

File: src/main/java/carpetextra/mixins/FlowerPotBlockMixin.java [hunk 5]
```diff
@@ -40,7 +36,7 @@ public FlowerPotBlockMixin(Settings settings)
         ),
         locals = LocalCapture.CAPTURE_FAILHARD
     )
-    private void onActivate(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockHitResult hit, CallbackInfoReturnable<ActionResult> cir, ItemStack itemStack, Item i, BlockState blockState) {
+    private void onActivate(BlockState blockState, World world, BlockPos pos, PlayerEntity player, BlockHitResult hit, CallbackInfoReturnable<ActionResult> cir) {
```
Comment (MeeniMc) [inline src/main/java/carpetextra/mixins/FlowerPotBlockMixin.java:39]: half broken: this works only when activated from a dispenser (with dispenserspotflowers ON), not when planting a rose by hand in the pot.

File: src/main/java/carpetextra/dispenser/behaviors/BlazePowderDispenserBehavior.java [hunk 6]
```diff
@@ -1,10 +1,6 @@
 package carpetextra.dispenser.behaviors;
 
-import net.minecraft.block.Block;
-import net.minecraft.block.BlockState;
-import net.minecraft.block.Blocks;
-import net.minecraft.block.DispenserBlock;
-import net.minecraft.block.NetherWartBlock;
+import net.minecraft.block.*;
```
[the remark this answers is not in the document]
Comment (voidpointer0x00) [reply src/main/java/carpetextra/dispenser/behaviors/BlazePowderDispenserBehavior.java:3]: intellij auto import optimize thingy :man_shrugging:

**Comment by MeeniMc** [comment]
### Summary of tested things so far: 
1. ✅ blazeMeal (hand, dispensed)
2. ✅ chickenShearing (hand, dispensed)
3. ✅ dispensersPlacesBlocks (all blocks that had a new 'dispenserBehavior' were tested, including heads, chests with items inside (picked chests))
4. ❓ dispensersPlaceBlocks with Carpets ON: pink carpets place but no debug info produced
5. ✅ dispensersCarvePumpkins
6. ✅ dispensersFeedAnimals (breed, grow)
7. ✅ dispensersFillMinecarts (no mixin changes in this PR)
8. ✅ dispensersMilkAnimals
9. ✅ dispensersPotPlants (normal, flowerPotChunkLoading)
10. ✅ dispensersStripBlocks
11. ✅ dispensersTillSoil
12. ❌ dispensersToggleThings
13. ❌ dispensersUseCauldrons (see details in reviews, filling empty bottles works, bucket works, rest doesn't)
14. ❌ flowerPutChunkLoading (by hand is broken, dispensed with dispensersPotPlants works)
15. ✅ renewableWitherSkeletons

### Not tested yet:
1. AutocraftingDropper (there is a mixin changed), but I don't know how proper behavior is for this modded behavior

File: src/main/java/carpetextra/dispenser/behaviors/CauldronWaterDispenserBehavior.java [hunk 7]
```diff
@@ -113,22 +108,24 @@ else if(frontBlock == Blocks.CAULDRON && item == Items.POTION && PotionUtil.getP
...
-        return item == Items.GLASS_BOTTLE ||
-            // water bottle
-            (item == Items.POTION && PotionUtil.getPotion(stack) == Potions.WATER) ||
-            // shulker boxes
-            Block.getBlockFromItem(item) instanceof ShulkerBoxBlock ||
-            // banners
-            item instanceof BannerItem ||
-            // dyeable items (leather armor, leather horse armor)
-            (item instanceof DyeableItem && ((DyeableItem) item).hasColor(stack));
+        if (item == Items.GLASS_BOTTLE || item instanceof BannerItem)
```
[the remark this answers is not in the document]
Comment (MeeniMc) [reply src/main/java/carpetextra/dispenser/behaviors/CauldronWaterDispenserBehavior.java:120]: I have fixes for all cauldron behaviors I'll pr on your branch

File: src/main/java/carpetextra/mixins/DispenserBehaviorCarpetsMixin.java (hunk 3 above)
Comment (MeeniMc) [reply src/main/java/carpetextra/mixins/DispenserBehaviorCarpetsMixin.java:12]: This is not a new defect (if it is even a defect).

File: src/main/java/carpetextra/mixins/FlowerPotBlockMixin.java (hunk 5 above)
Comment (MeeniMc) [reply src/main/java/carpetextra/mixins/FlowerPotBlockMixin.java:39]: https://github.com/voidpointer0x00/carpet-extra/pull/2

File: src/main/java/carpetextra/dispenser/behaviors/ToggleBlockDispenserBehavior.java (hunk 4 above)
Comment (MeeniMc) [reply src/main/java/carpetextra/dispenser/behaviors/ToggleBlockDispenserBehavior.java:55]: https://github.com/voidpointer0x00/carpet-extra/pull/2
I also added the new wood type buttons

**Comment by voidpointer0x00** [comment]
> ### Not tested yet:
> 1. AutocraftingDropper (there is a mixin changed), but I don't know how proper behavior is for this modded behavior

`autoCraftingDropper` looks to be working, tested the changed crafting part and filling with hopper slot by slot with comparator output.

**Comment by MeeniMc** [comment]
Looks like this is working and ready for review @altrisi

**Comment by nickmahdavi** [comment]
Not sure if this is in scope, but it seems like `emptyShulkerBoxStackAlways` isn't working properly server-side with Tweakeroo client-side. Empty shulkers get unstacked instantly. Seems like this was happening in 1.20.4 as well #316. Let me know if you need further testing or can't reproduce

**Comment by NickIndustries** [comment]
emptyShulkerBoxStackAlways is not working for me using version 1.4.141

**Comment by voidpointer0x00** [comment]
> emptyShulkerBoxStackAlways is not working for me using version 1.4.141

It seems if you only use `emptyShulkerBoxStackAlways` - it conflicts with carpet's `stackableShulkerBoxes` option.
I would encourage you to use the original carpet's `/carpet stackableShulkerBoxes` option as it stack only empty shulker boxes, not sure why `emptyShulkerBoxStackAlways` exists :smile: 

Although I found a bug with this update while testing - the option allows for non-empty shulkerboxes to stack :grin:

**Comment by voidpointer0x00** [comment]
@NickIndustries [this](https://github.com/voidpointer0x00/carpet-extra/releases/tag/1.20.6-pre4) release should have the `emptyShulkerBoxStackAlways` implementation corrected, although it's still preffered to use just the `stackableShulkerBoxes` option

**Comment by MeeniMc** [comment]
Good catch, did you try stacking boxes that have been placed and recollected with boxes that are virgin? In prior versions they may have non-empty NBT (that is, an NBT array with 0 elements, for example, rather than empty NBT).

**Comment by voidpointer0x00** [comment]
> did you try stacking boxes that have been placed and recollected with boxes that are virgin?

Yep, that works now

> StackableAlways also stack boxes in hoppers and other automated inventories

Tried that on the latest build, seemed to be working without the rule as well, though checking it right now with the old carpet & carpet extra (the first build for 1.20.6) wouldn't render the number of shulkers in other inventories, while there were still 2 of them on the server _(so it seems to be just a visual glitch)_.

**Comment by voidpointer0x00** [comment]
> just a few stylistic/reorders that I can fix easily enough afterwards or right before merging

I'll update the 1.21 port :)

Yeah, the style is a mess, could not figure out what's the projects default, to be honest :smile:

**Comment by voidpointer0x00** [comment]
> Seems like I filled that with conflicts with the last commits, sorry, I can try to resolve them/cherry pick them on top of new master if you want.

Reffered to that in the 1.21 pull request, I think I'll just hard reset and cherry pick it myself in a few hours _(working rn)_

**Comment by voidpointer0x00** [comment]
@altrisi https://github.com/gnembon/carpet-extra/issues/314 is now fixed, thanks for merging! :)

**Pull request merged**